ThreadPoolAllocator: Add support for updating the size of managed data

neobrain opened the pull request:
This will be used in #4479 to manage temporary buffers for parallel compilation.

neobrain commented on FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h:
```diff
@@ -427,13 +427,20 @@ class FixedSizePooledAllocation final {
-  Type ReownOrClaimBuffer() {
+  Type ReownOrClaimBuffer(std::optional<size_t> NewSize = std::nullopt) {
+    Size = NewSize.value_or(Size);
+    if (NewSize && ClientOwnedFlag != FEXCore::Utils::IntrusivePooledAllocator::ClientFlags::FLAG_FREE && (*Info)->Size < NewSize.value()) {
```
@Sonicadvance1 Is reading `(*Info)->Size` thread-safe for claimed-but-not-owned buffers? If not, we would have to move this check past ReownOrClaimBuffer and re-try if we find the size is insufficient.

Sonicadvance1 replied:
It's not safe.

neobrain replied:
Fixed - we'll now first attempt to reown the buffer and check its size, and then disclaim it and claim a new one if needed.

Sonicadvance1 commented:
Instead of hacking in support for dynamic sized buffer allocations in to the object that is supposed to only do fixed size allocations, hence the `FixedSizePooledAllocation` name. Either create a new wrapper that supports dynamically allocating a sized buffer or use the `IntrusivePooledAllocator` directly which understands handling pooled buffers of multiple different sizes.

neobrain commented:
> Instead of hacking in support for dynamic sized buffer allocations in to the object that is supposed to only do fixed size allocations, hence the FixedSizePooledAllocation name. Either create a new wrapper that supports dynamically allocating a sized buffer or use the IntrusivePooledAllocator directly which understands handling pooled buffers of multiple different sizes.

What do you find hacky about it? The implementation has no requirement for the size to be completely fixed as far as I can see. Perhaps the interface could use a new name, but an allocator optimized for high-frequency allocations of a buffer with rarely changing size seems like a valid use case to me. Copy-pasting the entire interface to do the same thing line-by-line but with one added feature doesn't sound like an improvement.

If it's the `std::optional` parameter that's bothering you, this could be moved to another overload of ReownOrClaimBuffer instead; or we could drop the `Size` from the constructor (putting responsibility on the caller not to change the value too often).

neobrain commented:
Fixed the issue about unsafe reading of metadata of a non-owned buffer. I've also renamed the helper class to better reflect that its main feature is the timer-based unclaiming of the managed buffer.
